File: src/js/main/commands/RichJson_clone.js
/*
    makes a deep copy of the given struct
*/

import {cloneObject} from "../helper/RichJsonHelper.js";
import {__RICH_JSON_CONFIG} from "../other/RichJsonConfiguration.js";
import {__RICH_JSON_COMMAND_PATH_DELIMITER} from "../core/RichJson.js";

export function __executeCloneCommand(parser, context) {
    if (parser.__isCloneApplying()) {
        if (__RICH_JSON_CONFIG.crashOnNestedCloneEnabled) {
            throw (`${parser.label} nested clone detected at '${context.currentPath.join(__RICH_JSON_COMMAND_PATH_DELIMITER)}'.`);
        }
        return currentMember;
    }
    parser.cache.cloneAddress = context.currentAddress;
    context.currentMember = cloneObject(context.currentMember);
    if (__RICH_JSON_CONFIG.debugEnabled) {
        console.debug(`${parser.label} resolved clone at '${context.currentPath.join(__RICH_JSON_COMMAND_PATH_DELIMITER)}'.`);
    }
    return context.currentMember;
}

File: src/python/main/commands/rich_json_clone.py
"""
Makes a deep copy of the given struct
"""
import logging

from ..helper.rich_json_helper import clone_object
from ..other.rich_json_configuration import _RICH_JSON_CONFIG
from ..core.rich_json_constants import _RICH_JSON_COMMAND_PATH_DELIMITER


def _execute_clone_command(parser, context):
    if parser._is_clone_applying():
        if _RICH_JSON_CONFIG["crashOnNestedCloneEnabled"]:
            raise ValueError(f"RichJson nested clone detected at '{_RICH_JSON_COMMAND_PATH_DELIMITER.join(map(str, context.current_path))}'.")
        return context.current_member

    parser.cache.clone_address = context.current_address
    context.current_member = clone_object(context.current_member)

    parser.logger.debug(f"RichJson resolved clone at '{_RICH_JSON_COMMAND_PATH_DELIMITER.join(map(str, context.current_path))}'.")

    return context.current_member